dask/array/slicing.py

```diff
@@ -231,10 +231,10 @@ def slice_wrap_lists(out_name, in_name, blockdims, index, itemsize):
     if not len(blockdims) == len(index):
         raise IndexError("Too many indices for array")
 
+    # Handle 0-d arrays, e.g. np.array(0)
+    index = [int(ind) if is_arraylike(ind) and ind.ndim == 0 else ind for ind in index]
```

Comment:
Since we're using `int()` to convert 0-dimensional NumPy arrays to Python scalars, in the case that someone passes in a non-integer NumPy scalar, this will result in us silently converting to a Python integer. With NumPy, this would raise an error instead:

```python
In [1]: import numpy as np

In [2]: x = np.array([4, 3, 2, 6, 4])

In [3]: x[np.array(1)]
Out[3]: 3

In [4]: x[np.array(1.2)]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-9943650c6805> in <module>
----> 1 x[np.array(1.2)]

IndexError: arrays used as indices must be of integer (or boolean) type
```

It would be good for us to fail loudly in this case instead of silently carrying on.

What happens if we use something like [`ndarray.item()`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.item.html) instead of `int()`? Does that raise an informative error to users?

Reply:
Dask already sanitizes given index when slicing so it will not be silently carrying on with an index like `np.array(1.2)`. Still, `ndarray.item()` is the correct way to do it for 0-dim NumPy array containing bool (e.g. `np.array(True)`) and I will be using that instead of `int()`. Thanks for the pointer!

```python
In [1]: import numpy as np

In [2]: x = np.array([4, 3, 2, 6, 4])

In [3]: x[np.array(1)]
Out[3]: 3

In [4]: x[np.array(1.2)]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-9943650c6805> in <module>
----> 1 x[np.array(1.2)]

IndexError: arrays used as indices must be of integer (or boolean) type

In [5]: import dask.array as da

In [6]: x_da = da.from_array(x)

In [7]: x_da[np.array(1)].compute()
Out[7]: 3

In [8]: x_da[np.array(1.2)].compute()
# IndexError: Bad index.  Must be integer-like: 1.2
```
<details>
   <summary>Details:</summary>
   
```python
In [1]: import numpy as np

In [2]: x = np.array([4, 3, 2, 6, 4])

In [3]: x[np.array(1)]
Out[3]: 3

In [4]: x[np.array(1.2)]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-9943650c6805> in <module>
----> 1 x[np.array(1.2)]

IndexError: arrays used as indices must be of integer (or boolean) type

In [5]: import dask.array as da

In [6]: x_da = da.from_array(x)

In [7]: x_da[np.array(1)].compute()
Out[7]: 3

In [8]: x_da[np.array(1.2)].compute()
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-8-0c4e157764d3> in <module>
----> 1 x_da[np.array(1.2)].compute()

~/Desktop/furkan/code/oss/dask/dask/array/core.py in __getitem__(self, index)
   1753         )
   1754 
-> 1755         index2 = normalize_index(index, self.shape)
   1756         dependencies = {self.name}
   1757         for i in index2:

~/Desktop/furkan/code/oss/dask/dask/array/slicing.py in normalize_index(idx, shape)
    911         if d is not None:
    912             check_index(i, d)
--> 913     idx = tuple(map(sanitize_index, idx))
    914     idx = tuple(map(normalize_slice, idx, none_shape))
    915     idx = posify_index(none_shape, idx)

~/Desktop/furkan/code/oss/dask/dask/array/slicing.py in sanitize_index(ind)
     89             check_int = np.isclose(index_array, int_index)
     90             first_err = index_array.ravel()[np.flatnonzero(~check_int)[0]]
---> 91             raise IndexError("Bad index.  Must be integer-like: %s" % first_err)
     92     else:
     93         raise TypeError("Invalid index type", type(ind), ind)

IndexError: Bad index.  Must be integer-like: 1.2
```
</details>

<br>

There are few other deviations from NumPy that I noticed when slicing Dask arrays. Do you think these should be addressed? If yes, here or in another PR?
<details>
  <summary>Other deviations</summary> 

```python
In [1]: import numpy as np

In [2]: import dask.array as da

In [3]: x = np.array([4, 3, 2, 6, 4])

In [4]: x_da = da.from_array(x)

In [5]: x[np.array("a")]  # informative error message
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-5-9833dd441084> in <module>
----> 1 x[np.array("a")]

IndexError: arrays used as indices must be of integer (or boolean) type

In [6]: x_da[np.array("a")].compute()  # not the most helpful error message 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-36c15aeea44f> in <module>
----> 1 x_da[np.array("a")].compute()

~/Desktop/furkan/code/oss/dask/dask/array/core.py in __getitem__(self, index)
   1753         )
   1754 
-> 1755         index2 = normalize_index(index, self.shape)
   1756         dependencies = {self.name}
   1757         for i in index2:

~/Desktop/furkan/code/oss/dask/dask/array/slicing.py in normalize_index(idx, shape)
    910     for i, d in zip(idx, none_shape):
    911         if d is not None:
--> 912             check_index(i, d)
    913     idx = tuple(map(sanitize_index, idx))
    914     idx = tuple(map(normalize_slice, idx, none_shape))

~/Desktop/furkan/code/oss/dask/dask/array/slicing.py in check_index(ind, dimension)
    972                     % (ind.size, dimension)
    973                 )
--> 974         elif (ind >= dimension).any() or (ind < -dimension).any():
    975             raise IndexError("Index out of bounds %s" % dimension)
    976     elif isinstance(ind, slice):

TypeError: '>=' not supported between instances of 'numpy.ndarray' and 'int'

In [7]: x[np.array(True)]  # returns x in an array
Out[7]: array([[4, 3, 2, 6, 4]])

In [8]: x[np.array(False)]  # returns empty array with shape (0, *x.shape)
Out[8]: array([], shape=(0, 5), dtype=int64)

In [9]: x_da[np.array(True)].compute()  # IndexError
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-9-662a9e3582d3> in <module>
----> 1 x_da[np.array(True)].compute()

~/Desktop/furkan/code/oss/dask/dask/array/core.py in __getitem__(self, index)
   1753         )
   1754 
-> 1755         index2 = normalize_index(index, self.shape)
   1756         dependencies = {self.name}
   1757         for i in index2:

~/Desktop/furkan/code/oss/dask/dask/array/slicing.py in normalize_index(idx, shape)
    910     for i, d in zip(idx, none_shape):
    911         if d is not None:
--> 912             check_index(i, d)
    913     idx = tuple(map(sanitize_index, idx))
    914     idx = tuple(map(normalize_slice, idx, none_shape))

~/Desktop/furkan/code/oss/dask/dask/array/slicing.py in check_index(ind, dimension)
    968         if ind.dtype == bool:
    969             if ind.size != dimension:
--> 970                 raise IndexError(
    971                     "Boolean array length %s doesn't equal dimension %s"
    972                     % (ind.size, dimension)

IndexError: Boolean array length 1 doesn't equal dimension 5

In [10]: x_da[np.array(False)].compute()  # IndexError
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-10-88af769857f6> in <module>
----> 1 x_da[np.array(False)].compute()

~/Desktop/furkan/code/oss/dask/dask/array/core.py in __getitem__(self, index)
   1753         )
   1754 
-> 1755         index2 = normalize_index(index, self.shape)
   1756         dependencies = {self.name}
   1757         for i in index2:

~/Desktop/furkan/code/oss/dask/dask/array/slicing.py in normalize_index(idx, shape)
    910     for i, d in zip(idx, none_shape):
    911         if d is not None:
--> 912             check_index(i, d)
    913     idx = tuple(map(sanitize_index, idx))
    914     idx = tuple(map(normalize_slice, idx, none_shape))

~/Desktop/furkan/code/oss/dask/dask/array/slicing.py in check_index(ind, dimension)
    968         if ind.dtype == bool:
    969             if ind.size != dimension:
--> 970                 raise IndexError(
    971                     "Boolean array length %s doesn't equal dimension %s"
    972                     % (ind.size, dimension)

IndexError: Boolean array length 1 doesn't equal dimension 5
```
</details>